src/pudl/analysis/timeseries_cleaning.py

```diff
@@ -1560,6 +1569,200 @@ def filter_missing_values(
     return df
 
 
+@dataclass
+class SimulateFlagsSettings:
+    """Define settings used to simulate flagged values for scoring imputation."""
+
+    num_months: int = 30
+    """The number of months of data to simulate."""
+    min_flag_rate: float = 0.1
+    """Min ratio of bad points in a section of data to be used for reference."""
+    max_flag_rate: float = 0.5
+    """Max ratio of bad points in a section of data to be used for reference."""
+
+
+class SimulationDataFrame(pa.DataFrameModel):
+    """Collect months of data which will be used to simulate flagged values.
+
+    Each row in this dataframe will contain a reference ID and month, and a simulation
+    ID and month. The reference ID/month points to a month of reported data where
+    a high ratio of values were flagged for imputation, and the simulation ID/month
+    points to a month of reported data with no flagged values. The flagged values
+    in the reference month will be used to simulate flagged values in the simulation
+    month.
+    """
+
+    reference_id_col: Series[Any]
+    reference_month: Series[pa.dtypes.DateTime]
+    simulation_id_col: Series[Any]
+    simulation_month: Series[pa.dtypes.DateTime]
+
+
+@pa.check_types
+def _merge_imputed(
+    aligned_df: DataFrame[AlignedTimeseriesDataFrame],
+    matrix: DataFrame[TimeseriesMatrix],
+    flags: DataFrame[TimeseriesMatrix],
+) -> pd.DataFrame:
+    """Helper function to melt imputed timeseries matrix and merge back on input asset."""
+    # Melt imputed matrix/flag matrix to long format
+    imputed_df = melt_imputed_timeseries_matrix(matrix, flags)
+
+    # Merge back on core table
+    return aligned_df.merge(
+        imputed_df.rename(columns={"value_col": "imputed_value_col"}),
+        on=["id_col", "datetime"],
+        how="left",
```

Comment:
Please add a merge validation for readability and error checking.

Reply:
Done.